### PR TITLE
GH-4037 fix for removing from VarNameCollector

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #4037 

Briefly describe the changes proposed in this PR:
 - optimized the code
 - create a new HashSet if we need to modify the set produced by the VarNameCollector

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

**Comment by hmottestad** [comment]
Still needs tests.

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/optimizer/QueryModelNormalizerOptimizer.java [hunk 1]
```diff
@@ -231,13 +231,40 @@ private boolean isWellDesigned(LeftJoin leftJoin) {
...
+					if (problemVars.contains(leftBindingName)) {
+						HashSet<String> ret = new HashSet<>(problemVars);
+						ret.removeAll(leftBindingNames);
+						return ret;
+					}
+				}
+			}
+		}
+		return problemVars;
+	}
```
Comment (hmottestad) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/optimizer/QueryModelNormalizerOptimizer.java:265]: This is the fix for the reported bug. I've also taken the liberty to check other places where we are assuming that a returned set is modifiable (which I feel is a bad assumption).

This code is covered by the tests in the SPARQL compliance module.

**Comment by hmottestad** [comment]
@maxstolze does this fix work for you? I wasn't able to reproduce the issue with the query and data you provided, but the stack trace makes me confident that this PR should fix the issue.

**Comment by maxstolze** [comment]
As far as I could test it, it fixes the issue. Thank you!

**Review by jeenbroekstra — APPROVED** [review]
Looks good to me!

**Pull request merged**